File: apps/patient/lib/core/outage/patient_readiness.dart
enum PatientReadinessRouteKind { public, internal }

enum PatientReadinessState { endpointUnverified, databaseUnavailable }

class PatientReadinessConfig {
  PatientReadinessConfig._();

  static const String path = '/health/patient-readiness';
  static const String endpointId = 'vhhealth-api';
  static const String purpose = 'patient_outage';
  static const int contractVersion = 1;
}

/// Strict, low-information response from the authenticated Patient readiness
/// route.
class PatientReadiness {
  static const Set<String> _allowedKeys = {
    'readinessContractVersion',
    'readinessPurpose',
    'ready',
    'state',
    'endpointId',
    'routeKind',
    'tenantId',
    'database',
    'serverTime',
  };

  const PatientReadiness({
    required this.ready,
    required this.state,
    required this.routeKind,
    required this.tenantId,
    required this.serverTime,
  });

  final bool ready;
  final PatientReadinessState? state;
  final PatientReadinessRouteKind? routeKind;
  final String? tenantId;
  final DateTime serverTime;

  factory PatientReadiness.fromJson(Map<String, dynamic> json) {
    final keys = json.keys.toSet();
    final unknownKeys = keys.difference(_allowedKeys);
    if (unknownKeys.isNotEmpty) {
      throw FormatException(
        'Unexpected patient-readiness fields: ${unknownKeys.join(', ')}',
      );
    }

    final contractVersion = json['readinessContractVersion'];
    final purpose = json['readinessPurpose'];
    final ready = json['ready'];
    if (contractVersion != PatientReadinessConfig.contractVersion ||
        purpose != PatientReadinessConfig.purpose ||
        ready is! bool) {
      throw const FormatException('Malformed patient-readiness response');
    }

    final expectedKeys = ready
        ? <String>{
            'readinessContractVersion',
            'readinessPurpose',
            'ready',
            'endpointId',
            'routeKind',
            'tenantId',
            'database',
            'serverTime',
          }
        : <String>{
            'readinessContractVersion',
            'readinessPurpose',
            'ready',
            'state',
            if (json.containsKey('routeKind')) 'routeKind',
            'serverTime',
          };
    if (keys.difference(expectedKeys).isNotEmpty ||
        expectedKeys.difference(keys).isNotEmpty) {
      throw const FormatException('Malformed patient-readiness field set');
    }

    final serverTimeRaw = json['serverTime'];
    if (serverTimeRaw is! String) {
      throw const FormatException('Malformed patient-readiness response');
    }
    final serverTime = DateTime.tryParse(serverTimeRaw);
    if (serverTime == null || !serverTime.isUtc) {
      throw const FormatException('Patient readiness serverTime must be UTC');
    }

    if (!ready) {
      return PatientReadiness(
        ready: false,
        state: _parseState(json['state']),
        routeKind: _parseRouteKind(json['routeKind']),
        tenantId: null,
        serverTime: serverTime,
      );
    }

    final tenantId = json['tenantId'];
    if (json['endpointId'] != PatientReadinessConfig.endpointId ||
        tenantId is! String ||
        tenantId.isEmpty ||
        json['database'] != 'ready') {
      throw const FormatException('Malformed ready response');
    }

    return PatientReadiness(
      ready: true,
      state: null,
      routeKind: _parseRequiredRouteKind(json['routeKind']),
      tenantId: tenantId,
      serverTime: serverTime,
    );
  }

  bool isReadyForTenant(String expectedTenantId) =>
      ready && tenantId == expectedTenantId && routeKind != null;

  static PatientReadinessRouteKind _parseRequiredRouteKind(Object? value) {
    final routeKind = _parseRouteKind(value);
    if (routeKind == null) {
      throw const FormatException('Missing patient-readiness routeKind');
    }
    return routeKind;
  }

  static PatientReadinessRouteKind? _parseRouteKind(Object? value) {
    return switch (value) {
      'public' => PatientReadinessRouteKind.public,
      'internal' => PatientReadinessRouteKind.internal,
      null => null,
      _ => throw const FormatException('Unknown patient-readiness routeKind'),
    };
  }

  static PatientReadinessState _parseState(Object? value) {
    return switch (value) {
      'endpoint_unverified' => PatientReadinessState.endpointUnverified,
      'database_unavailable' => PatientReadinessState.databaseUnavailable,
      _ => throw const FormatException('Unknown patient-readiness state'),
    };
  }
}

File: apps/backend/scripts/openapi/schemas/patientReadiness.mjs
import { envelope } from './_helpers.mjs';

export const schemas = {
  PatientReadiness: {
    type: 'object',
    additionalProperties: false,
    required: [
      'readinessContractVersion',
      'readinessPurpose',
      'ready',
      'endpointId',
      'routeKind',
      'tenantId',
      'database',
      'serverTime',
    ],
    properties: {
      readinessContractVersion: { type: 'integer', enum: [1] },
      readinessPurpose: { type: 'string', enum: ['patient_outage'] },
      ready: { type: 'boolean', enum: [true] },
      endpointId: { type: 'string', enum: ['vhhealth-api'] },
      routeKind: { type: 'string', enum: ['public', 'internal'] },
      tenantId: { type: 'string', format: 'uuid' },
      database: { type: 'string', enum: ['ready'] },
      serverTime: { type: 'string', format: 'date-time' },
    },
  },
  PatientReadinessResponse: envelope('PatientReadiness'),
  PatientReadinessFailure: {
    type: 'object',
    additionalProperties: false,
    required: [
      'readinessContractVersion',
      'readinessPurpose',
      'ready',
      'serverTime',
      'state',
    ],
    properties: {
      readinessContractVersion: { type: 'integer', enum: [1] },
      readinessPurpose: { type: 'string', enum: ['patient_outage'] },
      ready: { type: 'boolean', enum: [false] },
      routeKind: { type: 'string', enum: ['public', 'internal'] },
      serverTime: { type: 'string', format: 'date-time' },
      state: {
        type: 'string',
        enum: ['endpoint_unverified', 'database_unavailable'],
      },
    },
  },
  PatientReadinessFailureResponse: {
    type: 'object',
    required: ['success', 'message', 'code', 'details'],
    properties: {
      success: { type: 'boolean', enum: [false] },
      message: { type: 'string' },
      code: { type: 'string', enum: ['PATIENT_NOT_READY'] },
      details: {
        type: 'object',
        additionalProperties: false,
        required: ['readiness'],
        properties: {
          readiness: { $ref: '#/components/schemas/PatientReadinessFailure' },
        },
      },
      requestId: { type: 'string' },
    },
  },
};

export const operations = {
  'GET /api/v1/health/patient-readiness': {
    summary: 'Confirm patient app operational readiness',
    description:
      'Patient-only, low-information proof of the trusted ingress route, resolved active tenant, primary database, and server time. This contract intentionally does not report clinical-continuity policy state.',
    response: 'PatientReadinessResponse',
    additionalResponses: {
      503: {
        description: 'Patient readiness could not be verified',
        content: {
          'application/json': {
            schema: { $ref: '#/components/schemas/PatientReadinessFailureResponse' },
          },
        },
      },
    },
  },
};
